### PR TITLE
Feat: Badge Generator

### PR DESCRIPTION
**Pull request by kiznick**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description

In PR: #1119 Badges have been added but there's no telling if this feature exists and there's no way to use it other than reading the PR, I'd like to make a Badge Generator to make it easier to use.

## Type of change

Please delete any options that are not relevant.

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)
image
image

### CODE REVIEW

**Comment by Zaid-maker** [comment]
That's pretty amazing that someone added that functionality ❤️

File: src/components/BadgeGeneratorDialog.vue [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+    computed: {
+        badgeURL() {
+            if (!this.monitor.id || !this.badge.type) {
+                return;
+            }
+            let badgeURL = this.$root.baseURL + "/api/badge/" + this.monitor.id + "/" + this.badge.type;
+
+            // eslint-disable-next-line camelcase
+            let parameter_list = {};
```
Comment (chakflying) [inline src/components/BadgeGeneratorDialog.vue:239]:
```suggestion
            let parameterList = {};
```

File: src/components/BadgeGeneratorDialog.vue [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+            for (let parameter of this.parameters[this.badge.type] || []) {
+                if (parameter === "duration" && this.badge.duration) {
+                    badgeURL += "/" + this.badge.duration;
+                    continue;
+                }
+
+                if (this.badge[parameter]) {
+                    // eslint-disable-next-line camelcase
+                    parameter_list[parameter] = this.badge[parameter];
```
Comment (chakflying) [inline src/components/BadgeGeneratorDialog.vue:249]:
```suggestion
                    parameterList[parameter] = this.badge[parameter];
```

File: src/components/BadgeGeneratorDialog.vue [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            }
+
+            for (let parameter of [ "label", "style", "value" ]) {
+                if (parameter === "style" && this.badge.style === "flat") {
+                    continue;
+                }
+
+                if (this.badge[parameter]) {
+                    // eslint-disable-next-line camelcase
+                    parameter_list[parameter] = this.badge[parameter];
```
Comment (chakflying) [inline src/components/BadgeGeneratorDialog.vue:260]:
```suggestion
                    parameterList[parameter] = this.badge[parameter];
```

File: src/components/BadgeGeneratorDialog.vue [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                }
+            }
+
+            if (Object.keys(parameter_list).length > 0) {
+                return badgeURL + "?" + new URLSearchParams(parameter_list);
+            }
```
Comment (chakflying) [inline src/components/BadgeGeneratorDialog.vue:266]:
```suggestion
            if (Object.keys(parameterList).length > 0) {
                return badgeURL + "?" + new URLSearchParams(parameterList);
            }
```

**Review by kiznick — COMMENTED** [review]
Thanks for fix it!

**Review by chakflying — APPROVED** [review]
I'm not a fan of the original icon-only toggle for the public link, it's not easy to understand. But at least this is in line with the current UI.

**Comment by Tokynet** [comment]
What is holding up this PR?

File: src/components/BadgeGeneratorDialog.vue [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                            <option value="status">status</option>
+                            <option value="uptime">uptime</option>
+                            <option value="ping">ping</option>
+                            <option value="avg-response">avg-response</option>
+                            <option value="cert-exp">cert-exp</option>
+                            <option value="response">response</option>
+                        </select>
+                    </div>
+
+                    <div v-if=" (parameters[badge.type || 'null'] || [] ).includes('duration') " class="mb-3">
```
Comment (CommanderStorm) [inline src/components/BadgeGeneratorDialog.vue:24]: This `v-if` statement seems very ugly (and used more than once).
Have you considered a computed property for this usecase?

**Comment by kiznick** [comment]
> > I'm not a fan of the original icon-only toggle for the public link, it's not easy to understand. But at least this is in line with the current UI.
> 
> Tested this pr, it is working. I agree with @chakflying. The icons in the list is a bit strange. I am thinking maybe we can make them into a 3dot button.

What if we create another modal and merge 2 current buttons to this modal like this?
I think it will be useful if there will be more functions later.
image
image

**Comment by louislam** [comment]
Yes, it would be better.

**Comment by louislam** [comment]
The behaviour of the model is a little bit strange. When I incidentally drag it or try to highlight some text inside, it will disappear.

**Comment by kiznick** [comment]
> The behaviour of the model is a little bit strange. When I incidentally drag it or try to highlight some text inside, it will disappear.

Fixed.

**Comment by chakflying** [comment]
I think we need to add the clickable link strings to `en.json` to be able to translate it.

**Comment by kiznick** [comment]
> I think we need to add the clickable link strings to `en.json` to be able to translate it.

I can add in `/src/lang/en.json` right?

**Comment by chakflying** [comment]
Yes.

**Comment by mbh-dev** [comment]
Apologies if I'm being dumb, but how do you access this option within Uptime Kuma? I've successfully upgraded to Version: 1.22.0-beta.0 but cannot find this awesome feature anywhere.

Thanks

**Comment by louislam** [comment]
> > > I'm not a fan of the original icon-only toggle for the public link, it's not easy to understand. But at least this is in line with the current UI.
> > 
> > 
> > Tested this pr, it is working. I agree with @chakflying. The icons in the list is a bit strange. I am thinking maybe we can make them into a 3dot button.
> 
> What if we create another modal and merge 2 current buttons to this modal like this? I think it will be useful if there will be more functions later. image image

@mbh-dev See these images.

**Comment by RobertCSternberg** [comment]
A couple of questions that I believe would be helpful for many after the recent update.  In the next release, some examples would be great or an instruction page on Github could be helpful also. 

Have not figured out the answer to this yet. 
image

I think this might be based on the interval of the monitor. 
image

**Comment by joebailey26** [comment]
> Apologies if I'm being dumb, but how do you access this option within Uptime Kuma? I've successfully upgraded to Version: 1.22.0-beta.0 but cannot find this awesome feature anywhere.
> 
> Thanks

I too can't see this feature on any page after upgrading to 1.22.0

**Comment by CommanderStorm** [comment]
> I too can't see this feature on any page after upgrading to 1.22.0

The Badge generator can be found on the status page

**Comment by joebailey26** [comment]
> > I too can't see this feature on any page after upgrading to 1.22.0
> 
> The Badge generator can be found on the status page

Screenshot 2023-06-28 at 5 15 21 pm

**Comment by chakflying** [comment]
Did you try clicking "Edit Status Page"?

**Comment by joebailey26** [comment]
> Did you try clicking "Edit Status Page"?

image

**Comment by chakflying** [comment]
The "Open Badge Generator" button can be seen by clicking the Gear icon to the right of your monitor "Galexia BOS".

**Comment by joebailey26** [comment]
> The "Open Badge Generator" button can be seen by clicking the Gear icon to the right of your monitor "Galexia BOS".

Brilliant thank you! I've got it now. That's very tucked away.

**Pull request merged**